### PR TITLE
Fixing #1722

### PR DESCRIPTION
**Pull request by sithhell**
This PR fixes the problem described in #1722.

In order to be able to send parcels in batches, we need to serialize them right before sending them. This serialization can't block the current sending thread as we unnecessarily hold back parcels which are to be send in this batch. This eventually leads to deadlocks in some situations.
As such a third serialization pass has been added which is able to wait on futures that aren't ready yet when `put_parcel(s)` is called. This might be futures that are to be serialized or when a GID exceeds its credits and a incref is needed.

### CODE REVIEW

File: hpx/plugins/parcelport/mpi/sender.hpp [hunk 1]
```diff
@@ -48,64 +52,49 @@ namespace hpx { namespace parcelset { namespace policies { namespace mpi
...
+//                   , parcels_sent
+//                 )
+//             );
+//
+//             if(!sender->send())
+//             {
+//                 boost::unique_lock<mutex_type> l(connections_mtx_);
+//                 connections_.push_back(std::move(sender));
+//             }
+//         }
```
Comment (hkaiser) [inline hpx/plugins/parcelport/mpi/sender.hpp]: Can this commented code block be removed?

File: hpx/runtime/naming/name.hpp [hunk 2]
```diff
@@ -818,6 +826,8 @@ namespace hpx { namespace naming
 
             boost::detail::atomic_count count_;
             id_type_management type_;
+
+            mutable std::list<naming::gid_type> new_gids_;
```
Comment (hkaiser) [inline hpx/runtime/naming/name.hpp]: I think this adds overheads (in terms of memory requirements) which is not acceptable. There will be probably 1% of all ids which require special handling during serialization, but all of them will have a `std::list<gid_type>` member. Also, this does not make sense conceptually, id_type_impl is derived from gid_type and also holds a list of those? That smells.

File: hpx/runtime/naming/name.hpp [hunk 3]
```diff
@@ -781,6 +784,11 @@ namespace hpx { namespace naming
               : count_(0), type_(unknown_deleter)
             {}
 
+//             ~id_type_impl()
+//             {
+//                 HPX_ASSERT(new_gids_.empty());
+//             }
+
```
Comment (hkaiser) [inline hpx/runtime/naming/name.hpp]: Is this code still required?

File: hpx/runtime/parcelset/parcelport.hpp [hunk 4]
```diff
@@ -357,7 +357,14 @@ namespace hpx { namespace parcelset
         hpx::applier::applier *applier_;
 
         /// The cache for pending parcels
-        typedef std::pair<std::deque<parcel>, std::deque<write_handler_type> >
+#if defined(HPX_INTEL_VERSION) && HPX_INTEL_VERSION < 1400
+        typedef std::pair<
+            boost::shared_ptr<std::vector<parcel>>
```
Comment (hkaiser) [inline hpx/runtime/parcelset/parcelport.hpp]: Stylistic nitpick: we usually separate separate `>>` in templates.

File: src/runtime/naming/name.cpp [hunk 5]
```diff
@@ -278,53 +296,68 @@ namespace hpx { namespace naming
...
+                            {
+                                HPX_ASSERT(
+                                    overflow_credit <= HPX_GLOBALCREDIT_INITIAL-1);
+                                util::unlock_guard<scoped_lock> ul(ll);
+                                agas::decref(new_gid, overflow_credit);
+                            }
+
+                            return new_gid;
+                        }
+                    );
```
Comment (hkaiser) [inline src/runtime/naming/name.cpp]: I'd suggest to extract this big lambda into a separate function.

File: src/runtime/naming/name.cpp [hunk 6]
```diff
@@ -278,53 +296,68 @@ namespace hpx { namespace naming
...
                     boost::int64_t new_credit = (HPX_GLOBALCREDIT_INITIAL - 1) * 2;
-                    agas::incref(new_gid, new_credit);
+                    new_gid = gid;
+                    HPX_ASSERT(new_gid != invalid_gid);
+                    return agas::incref_async(new_gid, new_credit).then(
+                        [&gid](future<boost::int64_t>&&) mutable -> gid_type
+                        {
+                            typedef gid_type::mutex_type::scoped_lock scoped_lock;
+                            scoped_lock ll(gid.get_mutex());
+                            gid_type new_gid = gid;
```
Comment (hkaiser) [inline src/runtime/naming/name.cpp]: Is it guaranteed that `gid` is still alive when the lambda is executed?

Comment (sithhell) [inline src/runtime/naming/name.cpp]: Am 20.08.2015 2:42 nachm. schrieb "Hartmut Kaiser" <notifications@github.com
>:
>
> In src/runtime/naming/name.cpp:
>
> > @@ -278,53 +296,68 @@ namespace hpx { namespace naming
> >
> >                      util::unlock_guard<scoped_lock> ul(l);
> >                      boost::int64_t new_credit =
(HPX_GLOBALCREDIT_INITIAL - 1) * 2;
> > -                    agas::incref(new_gid, new_credit);
> > +                    new_gid = gid;
> > +                    HPX_ASSERT(new_gid != invalid_gid);
> > +                    return agas::incref_async(new_gid,
new_credit).then(
> > +                        [&gid](future<boost::int64_t>&&) mutable ->
gid_type
> > +                        {
> > +                            typedef gid_type::mutex_type::scoped_lock
scoped_lock;
> > +                            scoped_lock ll(gid.get_mutex());
> > +                            gid_type new_gid = gid;
>
> Is it guaranteed that gid is still alive when the lambda is executed?

Yes, the gid is a reference to the one that is supposed to be serialized.

File: hpx/runtime/naming/name.hpp (hunk 2 above)
Comment (sithhell) [inline hpx/runtime/naming/name.hpp]: I agree. It was a bad decision. Will update the PR ASAP.

File: hpx/runtime/naming/name.hpp (hunk 3 above)
Comment (sithhell) [inline hpx/runtime/naming/name.hpp]: No.

File: hpx/plugins/parcelport/mpi/sender.hpp (hunk 1 above)
Comment (sithhell) [inline hpx/plugins/parcelport/mpi/sender.hpp]: Yes.

File: hpx/runtime/parcelset/parcelport.hpp (hunk 4 above)
Comment (sithhell) [inline hpx/runtime/parcelset/parcelport.hpp]: *nod*. I'll setup a clang-format configuration for that not to happen again.

**Comment by sithhell** [comment]
The review comments have been addressed.

**Comment by hkaiser** [comment]
LGTM! Thanks!

**Pull request merged**